File: python/26_reciprocal_cycles.py
def reciprocal_cycles(n: int) -> int:
    max_length: int = 0
    max_num: int = 0
    for i in range(2, n):
        c = cycle(i)
        if c > max_length:
            max_num, max_length = i, c
    return max_num

def cycle(den: int) -> int:
    remainders = {} # use dict for fast lookup of remainder and its first occurrence
    rem = 1
    count = 0
    while rem != 0:
        rem = rem * 10 % den
        if rem in remainders:
            # cycle length is current decimal place - last decimal place remainder was seen
            return count - remainders[rem]
        else: # save new remainder in dict
            remainders[rem] = count
            count += 1
    return 0

if __name__ == "__main__":
    print(reciprocal_cycles(1000))

File: java/_26_ReciprocalCycles.java
import java.util.*;

public class _26_ReciprocalCycles{
    
    public static int getCycleLength(int num){
        int remainder = 1;
        int currStep = 0;
        HashMap<Integer, Integer> lastSeen = new HashMap<>();
        
        // Convert the fraction into a decimal and continue
        // until either a repeating set of remainders is found
        // or the decimal terminates.
        while (remainder != 0){
            remainder = (remainder * 10) % num;
            
            // If the remainder has been encountered previously,
            // there is a cycle in the decimal representation.
            if (lastSeen.containsKey(remainder)){
                return currStep - lastSeen.get(remainder);
            }
            
            // Store the digits and the current position.
            lastSeen.put(remainder, currStep);
            currStep++;
        }
        return 0; // terminating decimal
    }
    
    public static void main(String[] args){
        
        int num = 0;
        int longestCycle = 0;
        
        for (int d = 1; d < 1000; d++){
            int cycle = getCycleLength(d);
            System.out.printf("d: %d\tlen: %d\n", d, cycle);
            if (cycle > longestCycle){
                longestCycle = cycle;
                num = d;
            }
        }
        System.out.println("The integer under 1000 with the longest recurring cycle in its reciprocal is " + num); // 983
    }
}